src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
```

Comment:
This code will most likely be removed as it is part of the "incorrect" FTS solution that I was working on. Which in turn is why I didn't want to spent time on a slightly cleverer solution. I really don't know which container is best for this job, knowing that we will only try to insert 16 elements. If you think `std::unordered_set` has a smaller setup cost and is better for this job, I'm fine with switching.